File: 21.merge-two-sorted-lists.cpp
/*
 * [21] Merge Two Sorted Lists
 *
 * https://leetcode.com/problems/merge-two-sorted-lists
 *
 * algorithms
 * Easy (39.00%)
 * Total Accepted:    242.3K
 * Total Submissions: 620.8K
 * Testcase Example:  '[]\n[]'
 *
 * Merge two sorted linked lists and return it as a new list. The new list
 * should be made by splicing together the nodes of the first two lists.
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
       ListNode *head = nullptr; 
	   ListNode *cur = nullptr;
	   while(l1 && l2){
		   ListNode *p = l1->val <= l2->val ? l1 : l2;
		   if(cur){
			   cur->next = p;
			   cur = p;
		   }
		   else{
			   head = cur = p;
		   }

		   if(p==l1){
			   l1 = l1->next;
		   }
		   else{
			   l2 = l2->next;
		   }

		   cur->next = nullptr;
	   }

	   if(l1 || l2){
		   ListNode* p = l1 ? l1 : l2;
		   if(cur){
			   cur->next = p;
			   cur = p;
		   }
		   else{
			   head = cur = p;
		   }
	   }

	   return head;
    }
};

File: 21.merge-two-sorted-lists.go
/*
 * @lc app=leetcode id=21 lang=golang
 *
 * [21] Merge Two Sorted Lists
 *
 * https://leetcode.com/problems/merge-two-sorted-lists/description/
 *
 * algorithms
 * Easy (47.66%)
 * Total Accepted:    606K
 * Total Submissions: 1.3M
 * Testcase Example:  '[1,2,4]\n[1,3,4]'
 *
 * Merge two sorted linked lists and return it as a new list. The new list
 * should be made by splicing together the nodes of the first two lists.
 * 
 * Example:
 * 
 * Input: 1->2->4, 1->3->4
 * Output: 1->1->2->3->4->4
 * 
 * 
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	var head, tail, curr *ListNode
	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			curr = l1
			l1 = l1.Next
		} else {
			curr = l2
			l2 = l2.Next
		}
		if tail != nil {
			tail.Next = curr
		} else {
			head = curr
		}
		curr.Next = nil
		tail = curr
	}
	if l1 != nil {
		curr = l1
	} else {
		curr = l2
	}
	if curr != nil {
		if tail != nil {
			tail.Next = curr
		} else {
			head = curr
		}
	}

	return head
}
